# HÆ°á»›ng dáº«n Viáº¿t Test Spec & Integration Test Chuáº©n

> TÃ i liá»‡u nÃ y ra Ä‘á»i sau sá»± cá»‘: *Declaration module Ä‘á»•i schema lá»›n nhÆ°ng Integration Test khÃ´ng Ä‘Æ°á»£c cáº­p nháº­t, dáº«n Ä‘áº¿n toÃ n bá»™ test file sá»­ dá»¥ng schema cÅ© vÃ  khÃ´ng thá»ƒ cháº¡y Ä‘Æ°á»£c.*

---

## ğŸ”´ Cháº©n Ä‘oÃ¡n: 3 Äiá»ƒm gÃ£y trong quy trÃ¬nh cÅ©

| # | Äiá»ƒm gÃ£y | Háº­u quáº£ |
|---|---|---|
| 1 | **Test Spec quÃ¡ sÆ¡ sÃ i** â€” Chá»‰ cÃ³ Scenario cáº¥p cao, khÃ´ng cÃ³ Input/Output cá»¥ thá»ƒ | AI/Dev khÃ´ng cÃ³ Ä‘á»§ ngá»¯ cáº£nh Ä‘á»ƒ viáº¿t test Ä‘Ãºng |
| 2 | **Test chÆ°a bao giá» Ä‘Æ°á»£c cháº¡y sau khi schema thay Ä‘á»•i** | Schema mismatch Ã¢m tháº§m tá»“n táº¡i, bá»‹ phÃ¡t hiá»‡n muá»™n |
| 3 | **KhÃ´ng cÃ³ convention rÃ ng buá»™c Test Spec â†” Test Code** | Test code "trÃ´i dáº¡t" khá»i spec, khÃ´ng ai biáº¿t TC nÃ o Ä‘Ã£ Ä‘Æ°á»£c cover |

---

## âœ… Giáº£i phÃ¡p: 3 Thay Ä‘á»•i Quy trÃ¬nh

---

### THAY Äá»”I 1: Test Spec pháº£i viáº¿t Ä‘áº¿n cáº¥p Input/Output (Granular Test Case)

#### âŒ Chuáº©n cÅ© (quÃ¡ sÆ¡ sÃ i â€” KHÃ”NG dÃ¹ng ná»¯a)
```
### Scenario 2: TÃ­nh toÃ¡n cÃ´ng thá»©c ngáº§m
- Test Case 2.1: Update Declaration API PUT /api/declarations/:id
  - Body Valid: invoicePriceBeforeVat = 100000, declarationQuantity = 10
  - Expect: totalLotValueBeforeVat = 1000000
```

#### âœ… Chuáº©n má»›i (Granular â€” Báº®T BUá»˜C Ã¡p dá»¥ng)

Má»—i Test Case trong file `.md` pháº£i cÃ³ Ä‘á»§ 4 pháº§n: **ID, Input, Expected Output, Ghi chÃº logic**.

```markdown
### TC-DECL-UPDATE-01: Happy Path â€” Cáº­p nháº­t thÃ nh cÃ´ng cÃ¡c trÆ°á»ng text

- **Endpoint**: `PUT /api/declarations/:id`
- **Auth**: ADMIN token
- **Precondition**: Declaration tá»“n táº¡i (id = X), Ä‘Æ°á»£c táº¡o qua ProductCode
- **Input Body**:
  ```json
  {
    "brand": "Nike",
    "sellerCompanyName": "Nike China Ltd",
    "declarationNeed": "Nháº­p kháº©u thÆ°Æ¡ng máº¡i",
    "notes": "Ghi chÃº test"
  }
  ```
- **Expected HTTP Status**: `200`
- **Expected Response**:
  ```json
  { "code": 200, "message": "Success", "data": { "brand": "Nike", ... } }
  ```
- **DB Verify**: `declaration.brand = "Nike"`, `declaration.sellerCompanyName = "Nike China Ltd"`

---

### TC-DECL-UPDATE-02: Secure Recalculation â€” totalLotValueBeforeVat

- **Endpoint**: `PUT /api/declarations/:id`
- **Auth**: ADMIN token
- **Input Body**:
  ```json
  {
    "invoicePriceBeforeVat": 100000,
    "declarationQuantity": 10,
    "importTax": 5,
    "vatTax": 10,
    "totalLotValueBeforeVat": 99000000
  }
  ```
  > âš ï¸ `totalLotValueBeforeVat: 99000000` lÃ  giÃ¡ trá»‹ giáº£ máº¡o tá»« Client
- **Expected HTTP Status**: `200`
- **DB Verify** (giÃ¡ trá»‹ thá»±c táº¿ Server tá»± tÃ­nh):
  - `totalLotValueBeforeVat = 100000 Ã— 10 = 1,000,000` (KHÃ”NG pháº£i 99,000,000)
  - `importTaxPayable = 1,000,000 Ã— 5% = 50,000`
  - `vatTaxPayable = 1,000,000 Ã— 10% = 100,000`
```

#### Checklist Test Spec Ä‘á»§ granular
Má»—i API Endpoint cáº§n cÃ³ Test Case cho:
- [ ] **Happy Path**: Input há»£p lá»‡ Ä‘áº§y Ä‘á»§ â†’ 200/201
- [ ] **Auth â€” No Token**: KhÃ´ng cÃ³ Bearer token â†’ 401
- [ ] **Auth â€” Wrong Role**: Token cá»§a role khÃ´ng Ä‘á»§ quyá»n â†’ 403
- [ ] **Not Found**: ID khÃ´ng tá»“n táº¡i â†’ 404
- [ ] **Validation â€” Missing Required**: Thiáº¿u field báº¯t buá»™c â†’ 400
- [ ] **Validation â€” Invalid Type**: Sai kiá»ƒu dá»¯ liá»‡u (string thay vÃ¬ number) â†’ 400
- [ ] **Business Logic**: Verify dá»¯ liá»‡u tÃ­nh toÃ¡n tá»± Ä‘á»™ng Ä‘Ãºng cÃ´ng thá»©c â†’ DB check
- [ ] **Cache**: Verify cache bá»‹ xÃ³a sau mutation (POST/PUT/DELETE) â†’ Redis check
- [ ] **Edge Case**: GiÃ¡ trá»‹ biÃªn (0, null, sá»‘ Ã¢m, ráº¥t lá»›n...)

---

### THAY Äá»”I 2: Convention "TC-ID Comment" rÃ ng buá»™c Test Spec â†” Test Code

#### Quy táº¯c báº¯t buá»™c
Má»—i `it()` block trong Integration Test pháº£i báº¯t Ä‘áº§u báº±ng comment chá»©a **Test Case ID** tÆ°Æ¡ng á»©ng tá»« Test Spec.

#### âœ… Chuáº©n má»›i
```javascript
describe('PUT /api/declarations/:id â€” Update', () => {

    // [TC-DECL-UPDATE-01] Happy Path â€” Cáº­p nháº­t thÃ nh cÃ´ng cÃ¡c trÆ°á»ng text
    it('should update text fields successfully', async () => { ... });

    // [TC-DECL-UPDATE-02] Secure Recalculation â€” cháº·n fake totalLotValueBeforeVat
    it('should recalculate totalLotValueBeforeVat server-side, ignoring client value', async () => { ... });

    // [TC-DECL-UPDATE-AUTH-01] Auth â€” Non-admin bá»‹ tá»« chá»‘i
    it('should return 403 when non-admin tries to update', async () => { ... });

    // [TC-DECL-UPDATE-404] Not Found â€” Declaration khÃ´ng tá»“n táº¡i
    it('should return 404 for non-existent declaration', async () => { ... });
});
```

**Lá»£i Ã­ch**:
- Scan nhanh xem TC nÃ o Ä‘Ã£ cÃ³ `it()` tÆ°Æ¡ng á»©ng, TC nÃ o bá»‹ bá» sÃ³t
- Khi Test Spec thay Ä‘á»•i, grep theo ID Ä‘á»ƒ tÃ¬m ngay test cáº§n update
- AI cÃ³ Ä‘á»§ ngá»¯ cáº£nh Ä‘á»ƒ hiá»ƒu **táº¡i sao** `it()` block Ä‘Ã³ tá»“n táº¡i

---

### THAY Äá»”I 3: Checklist báº¯t buá»™c trÆ°á»›c khi nÃ³i "Implementation Done"

Khi má»™t feature/module Ä‘Æ°á»£c coi lÃ  **Done**, pháº£i pass háº¿t checklist sau:

```
â˜ 1. BRD cáº­p nháº­t Ä‘áº§y Ä‘á»§
â˜ 2. Tech Spec cáº­p nháº­t Ä‘áº§y Ä‘á»§ (schema, API contract)
â˜ 3. Test Spec viáº¿t Ä‘á»§ granular (cÃ³ TC-ID, Input, Expected Output)
â˜ 4. Integration Test viáº¿t xong (má»—i TC trong Spec cÃ³ 1 it() block, cÃ³ TC-ID comment)
â˜ 5. Cháº¡y Integration Test: npm run test -- --testPathPattern=<tÃªn-module>.test.js
â˜ 6. Táº¥t cáº£ test PASS (mÃ u xanh 100%)
â˜ 7. KhÃ´ng cÃ³ SKIP test (khÃ´ng dÃ¹ng it.skip() Ä‘á»ƒ trá»‘n lá»—i)
```

> ğŸ”´ **Báº¥t ká»³ bÆ°á»›c nÃ o chÆ°a xong = Feature chÆ°a Done.**
> Tuyá»‡t Ä‘á»‘i khÃ´ng chuyá»ƒn sang lÃ m Frontend khi test chÆ°a xanh.

---

## ğŸ“ Template Test Spec chuáº©n

DÆ°á»›i Ä‘Ã¢y lÃ  template má»—i file TestSpec má»›i nÃªn sá»­ dá»¥ng:

```markdown
# Test Spec: Module [TÃªn Module] â€” Backend

> **PhiÃªn báº£n**: v1.0 | **Cáº­p nháº­t**: YYYY-MM-DD
> **Tham chiáº¿u**: [BRD link], [Tech Spec link]
> **Schema hiá»‡n táº¡i**: Liá»‡t kÃª cÃ¡c trÆ°á»ng model Prisma Ä‘ang dÃ¹ng (Ä‘á»ƒ trÃ¡nh mismatch)

## Danh sÃ¡ch API cáº§n test

| Endpoint | Method | Auth | MÃ´ táº£ |
|---|---|---|---|
| `/api/xxx` | GET | Any | Láº¥y danh sÃ¡ch |
| `/api/xxx/:id` | PUT | ADMIN | Cáº­p nháº­t |

---

## Test Cases

### [TC-XXX-GET-01] â€” GET danh sÃ¡ch: Happy Path
...

### [TC-XXX-GET-AUTH-01] â€” GET danh sÃ¡ch: KhÃ´ng cÃ³ token â†’ 401
...

### [TC-XXX-UPDATE-01] â€” PUT update: Happy Path
...
```

---

## ğŸ”„ Quy trÃ¬nh Change Management khi Schema thay Ä‘á»•i

Khi cÃ³ thay Ä‘á»•i schema (thÃªm/xÃ³a/Ä‘á»•i tÃªn field), báº¯t buá»™c thá»±c hiá»‡n theo thá»© tá»±:

```
1. Cáº­p nháº­t schema.prisma + cháº¡y migration
        â†“
2. Cáº­p nháº­t BRD (náº¿u thay Ä‘á»•i nghiá»‡p vá»¥)
        â†“
3. Cáº­p nháº­t Tech Spec (Data Dictionary)
        â†“
4. Cáº­p nháº­t Test Spec:
   - XÃ³a/sá»­a Test Case dÃ¹ng field cÅ©
   - ThÃªm Test Case cho field má»›i
        â†“
5. Cáº­p nháº­t Integration Test:
   - Sá»­a táº¥t cáº£ prisma.create/createMany Ä‘ang dÃ¹ng field cÅ©
   - ThÃªm it() block cho Test Case má»›i
        â†“
6. Cháº¡y láº¡i toÃ n bá»™ test suite cá»§a module Ä‘Ã³
   npm run test -- --testPathPattern=<module>.test.js
        â†“
7. 100% PASS â†’ Done
```

> âš ï¸ **Anti-pattern cáº§n trÃ¡nh**: Chá»‰ sá»­a code Controller/Service mÃ  KHÃ”NG cháº¡y láº¡i test.
> ÄÃ¢y chÃ­nh xÃ¡c lÃ  nguyÃªn nhÃ¢n táº¡o ra file test vá»›i schema cÅ©.

---

*TÃ i liá»‡u nÃ y lÃ  Rule báº¯t buá»™c, Ã¡p dá»¥ng cho cáº£ Engineers vÃ  AI Agent.*
